### PR TITLE
Gradle 6.7.1, AGP 4.2, robolectric 4.5.1, mockk 1.11.0, and other smaller upgrades

### PR DESCRIPTION
**Pull request by LukasPaczos**
### TL;DR
@mapbox/navigation-android there are incompatibilities between our tooling and JDK 11. There are also issues upgrading to the latest version of the tools that should resolve those incompatibilities. I'm opening this as a draft because I don't have the bandwidth to complete this right away, it became a huge rabbit hole.

To avoid the issues described below while you're building the `main` branch, just make sure that your local environment uses JDK 8 and not JDK 11 that Android Studio ships as default now.

### Description

Closes https://github.com/mapbox/mapbox-navigation-android/issues/3887.

#### Java 11 and Android Gradle Plugin 4.2
Android Studio 4.2+ now ships by default with Java 11, this means that in a default setup, Gradle also runs on Java 11 now which can various sorts of issues and incompatibilities. This PR resolves those as well as upgrades dependencies and cleans up the code. There's also available Android Gradle Plugin 4.2. To be able to upgrade to that plugin we need https://github.com/mapbox/android-sdk-versions-plugin/issues/29 and update to the versions plugin to `v1.1.1`.

I also added a recommended `-XX:+UseParallelGC` to speed up the builds with the new garbage collector.

Update to Android Gradle Plugin to `v4.2` also requires updating Gradel to `v6.7.1`.

#### robolectric and mockk
With the new Java version, Robolectric is running into issues with key stores (https://github.com/robolectric/robolectric/issues/5496) so it requires an update to `v4.4+` to resolve the issue. This upgrade started causing mockk to fail for some objects with an exception like:
```
java.lang.ClassCastException: Cannot cast com.mapbox.common.TileStore to com.mapbox.common.TileStore
```
which is caused by class loader duplication (https://github.com/robolectric/robolectric/issues/5848) introduced by mockk's incompatibility with https://github.com/robolectric/robolectric/pull/5036 and fixed in https://github.com/mockk/mockk/pull/472 (`v1.10.1+`).

However, there's also a similar cast exception thrown like this:
```
Caused by: java.lang.ClassCastException: class com.mapbox.navigator.RouterError cannot be cast to class com.mapbox.navigator.RouterError (com.mapbox.navigator.RouterError is in unnamed module of loader 'app'; com.mapbox.navigator.RouterError is in unnamed module of loader org.robolectric.internal.AndroidSandbox$SdkSandboxClassLoader @11bd0f3b)
```
which fits the above issue that should be resolved but there has to be somethign more to it. Exlcuding the package from the robolectric's sandbox mostly works around it for vulnerable classes but I still don't understand the root cause:
```kotlin
class NavSDKRobolectricTestRunner(testClass: Class<*>) : RobolectricTestRunner(testClass) {
    override fun createClassLoaderConfig(method: FrameworkMethod?): InstrumentationConfiguration {
        val builder = InstrumentationConfiguration.Builder(super.createClassLoaderConfig(method))
        builder.doNotAcquirePackage("com.mapbox.navigator")
        return builder.build()
    }
}
```

I occasionally still see `java.lang.LinkageError` which again would point to the same issue of multiple class loaders that should be resolved but isn't for me locally. Fully resolving this is still TBD.

There are also other smaller mockk incompatibilities that need to still be resolved. One of them is mockk reporting multiple verifications while there only is one in code. Could that multiple class loaders issue again?

Mockk now also allows for passing method references to mock extension functions! Instead of
```
mockkStatic("com.mapbox.navigation.base.internal.extensions.ContextEx")
```
we can now simply do:
```
mockkStatic(Context::inferDeviceLocale)
```

#### jacoco
Robolectric needs the `jacoco.includeNoLocationClasses = true` flag to correctly measure coverage for robolectric tests but on Java 9+ this results in exceptions thrown by picking up internal JDK classes. This is resolved with
```
jacoco.excludes = ['jdk.internal.*']
```

#### Other cleanups
I also removed unnecessary `@JvmName` annotations for internal APIs and added correct cleanups for tests that use `mockkStatic`.

### Changelog

```
<changelog>TBD</changelog>
```

### CODE REVIEW

**Comment by zugaldia** [comment]
What's the impact of not landing this PR, that developers wouldn't be able to use the Nav SDK with Android Studio 4.2+?

**Comment by LukasPaczos** [comment]
No, the impact is that:
- we can't build the SDK locally if our environment uses JDK 11
- we can't adopt features and fixes from the latest versions of AGP, robolectric, and mockk (going back as much as a year now)

Nothing blocking an actual SDK release, but might slow the project development down at some point if we hit a blocker and won't be able to upgrade. I didn't expect this to become such a huge effort, that's why postponing for now until there's a little bit more time to wrap this up.

**Comment by LukasPaczos** [comment]
Closing in favor of https://github.com/mapbox/mapbox-navigation-android/pull/5192.

This PR contains a little bit more changes that take advantage of the upgraded versions of tools (like parallel GC or mockk class refs for extensions), but we can reintroduce them separately.
cc @Zayankovsky
